test/test_sparse.py

```diff
@@ -4890,6 +4890,53 @@ def run_test(batch_shape, m, n, k, device, dtype, dtype_out, add_bias, activatio
             k = 2 ** k * 128
             run_test(batch_shape, m, n, k, device, dtype, dtype_out[dtype], add_bias, activation, rtol, atol)
 
+    @unittest.skipIf(TEST_WITH_ROCM, "ROCm doesn't support CUTLASS")
+    @dtypes(torch.int8, torch.half, torch.bfloat16)
+    def test_semi_structured_sparse_conversions(self, device, dtype):
+        from torch.sparse.semi_structured_sparse_conversions import (
+            semi_structured_sparse_compress_2by4,
+            semi_structured_sparse_uncompress_2by4
+        )
+
+        def make_tensor(shape, dtype):
+            if dtype.is_complex:
+                return torch.zeros(shape, dtype=dtype)
+            elif dtype.is_floating_point:
+                x = torch.randn(shape, dtype=dtype) / 10
+                x[x == 0] = 1
+                return x
+            else:
+                x = torch.randint(-5, 5, shape, dtype=dtype)
+                x[x == 0] = 1
+                return x
+
+        def random_mask_choice():
+            choices = [
+                [1, 1, 0, 0],
+                [1, 0, 1, 0],
+                [1, 0, 0, 1],
+                [0, 1, 1, 0],
+                [0, 1, 0, 1],
+                [0, 0, 1, 1]
+            ]
+            i = random.randint(0, len(choices) - 1)
+            return choices[i]
+
+        def run_test(m, k, device, dtype):
+            mask_entries = [random_mask_choice() for i in range(m * (k // 4))]
+            mask = torch.tensor(mask_entries, dtype=torch.bool).view(m, k)
+            dense_ref = make_tensor((m, k), dtype).masked_fill(~mask, 0).to(device)
+
+            sparse, meta = semi_structured_sparse_compress_2by4(dense_ref)
+            dense = semi_structured_sparse_uncompress_2by4(sparse, meta)
+
+            torch.testing.assert_close(dense, dense_ref, rtol=0, atol=0)
```

Comment:
This is a good test to have, but it would be good to have some kind of validation against the `sparse`, and `meta` products of the compression method. 

Testing only the roundtrip could lead to a state where changes to both routines break the compression and decompression in such a way that the round trip is correct but the compressed products are not.

Reply:
To add checking "compress" results against CUTLASS reference implementation:

<details>
<summary>
Save this code say as `aten/src/ATen/native/sparse/cuda/SemiStructuredSparseCompressCUTLASS.cu`
</summary>

```C++
#include <ATen/core/Tensor.h>

#include <torch/torch.h>

#ifndef USE_ROCM
#include <cuda_runtime.h>
#include <cutlass/cutlass.h>
#include <cutlass/layout/layout.h>
#include <cutlass/tensor_ref.h>
#endif

#include <tuple>

namespace at {
namespace native {

// Copied from tools/util/include/host_reorder.h, from CUTLASS source
// tree.  If this file included above, it would produce build errors,
// thus copying; the whole code in this file should be short-lived, as
// its sole purpose is testing of
// torch/sparse/semi_structured_sparse_compress.py code, so it should
// be OK this way.
template <typename Element, typename LayoutDest, typename LayoutSrc>
static void reorder_meta(cutlass::TensorRef<Element, LayoutDest> dest,
                         cutlass::TensorRef<Element, LayoutSrc> src,
                         const int problem_size_m, const int problem_size_k) {
  for (int m = 0; m < problem_size_m; m++) {
    for (int k = 0; k < problem_size_k; k++) {
      // First reorder the rows.
      int group = (sizeof(Element) == 2) ? 32 : 16;
      int interweave = (sizeof(Element) == 2) ? 4 : 2;

      int dest_row = m / group * group + (m % 8) * interweave + (m % group) / 8;
      int dest_col = k;

      // Next swizzle the 2x2 blocks from Z to N.
      if (((dest_row % 2) == 0) && ((dest_col % 2) == 1)) {
        ++dest_row;
        --dest_col;
      } else if (((dest_row % 2) == 1) && ((dest_col % 2) == 0)) {
        --dest_row;
        ++dest_col;
      }

      dest.at({dest_row, dest_col}) = src.at({m, k});
    }
  }
}

std::tuple<Tensor, Tensor>
_semi_structured_sparse_compress_2by4_cutlass(const Tensor& dense) {
#ifndef USE_ROCM
  // Check dimensions of the dense matrix.
  TORCH_CHECK(dense.dim() == 2,
              "_semi_structured_sparse_compress_2by4_cutlass: Expected dense "
              "argument to be 2D tensor, got ", dense.dim(), " dims");

  // Determine PyTorch datatype for the metadata matrix.
  auto meta_dtype = at::kChar;
  auto dense_elems_per_meta_elem = 0;
  if (dense.dtype() == at::kChar) {
    meta_dtype = at::kInt;
    dense_elems_per_meta_elem = 32;
  } else if (dense.dtype() == at::kHalf || dense.dtype() == at::kBFloat16) {
    meta_dtype = at::kShort;
    dense_elems_per_meta_elem = 16;
  } else {
    AT_ERROR("semi_structured_sparse_compress_2by4_cutlass: Invalid dense "
             "argument datatype ", dense.dtype(), "encountered");
  }

  const auto dense_nrows = dense.size(0);
  const auto dense_ncols = dense.size(1);

  if (dense_nrows % 32 != 0) {
    AT_ERROR("_semi_structured_sparse_compress_2by4_cutlass: Number of rows of"
             "dense matrix must be divisible by 32, but it is ",
             dense_nrows);
  }
  if (dense_ncols % dense_elems_per_meta_elem != 0) {
    AT_ERROR("_semi_structured_sparse_compress_2by4_cutlass: Number of columns"
             "of dense matrix must be divisible by ",
             dense_elems_per_meta_elem, ", but it is ", dense_ncols);
  }

  const auto dense_cpu = dense.to("cpu");

  const auto mask_cpu = dense_cpu != 0;

  const auto sparse_cpu =
    dense_cpu.masked_select(mask_cpu).view({dense_nrows, dense_ncols / 2});

  const auto meta_nrows = dense_nrows;
  const auto meta_ncols = dense_ncols / dense_elems_per_meta_elem;

  auto meta_cpu = dense_cpu.new_empty({meta_nrows, meta_ncols},
                                      at::TensorOptions().dtype(meta_dtype));
  auto* mask_cpu_ptr = mask_cpu.data_ptr<bool>();
  for (auto i = 0; i < meta_nrows; ++i) {
    for (auto j = 0; j < meta_ncols; ++j) {
      uint64_t meta_val = 0;
      for (auto k = 0; k < dense_elems_per_meta_elem / 4; ++k, mask_cpu_ptr += 4) {
        const auto mask_elems = std::make_tuple(
          mask_cpu_ptr[0],
          mask_cpu_ptr[1],
          mask_cpu_ptr[2],
          mask_cpu_ptr[3]
        );
        auto meta_quadruple = 0;
        if (mask_elems == std::make_tuple(1, 1, 0, 0)) {
          meta_quadruple = 4; // 0100
        } else if (mask_elems == std::make_tuple(1, 0, 1, 0)) {
          meta_quadruple = 8; // 1000
        } else if (mask_elems == std::make_tuple(0, 1, 1, 0)) {
          meta_quadruple = 9; // 1001
        } else if (mask_elems == std::make_tuple(1, 0, 0, 1)) {
          meta_quadruple = 12; // 1100
        } else if (mask_elems == std::make_tuple(0, 1, 0, 1)) {
          meta_quadruple = 13; // 1101
        } else if (mask_elems == std::make_tuple(0, 0, 1, 1)) {
          meta_quadruple = 14; // 1110
        }
        else {
          AT_ERROR("semi_structured_sparse_compress_2by4_cutlass: dense "
                   "argument does not match 2:4 sparsity pattern");
        }
        meta_val = meta_val | (meta_quadruple << (4 * k));
      }
      const auto idx = i * meta_ncols + j;
      if (meta_dtype == at::kShort) {
        using MetaElement = int16_t;
        const auto meta_cpu_ptr = meta_cpu.data_ptr<MetaElement>();
        meta_cpu_ptr[idx] = (MetaElement)meta_val;
      } else if (meta_dtype == at::kInt) {
        using MetaElement = int32_t;
        const auto meta_cpu_ptr = meta_cpu.data_ptr<MetaElement>();
        meta_cpu_ptr[idx] = (MetaElement)meta_val;
      }
    }
  }

  auto meta_reordered_cpu = meta_cpu.new_empty({meta_nrows, meta_ncols});
  using MetaLayout = cutlass::layout::RowMajor;
  using MetaReorderedLayout = cutlass::layout::ColumnMajorInterleaved<2>;
  if (meta_dtype == at::kShort) {
    using MetaElement = int16_t;
    auto meta_cpu_ref =
      cutlass::TensorRef<MetaElement, MetaLayout>(
          meta_cpu.data_ptr<MetaElement>(),
          MetaLayout::packed({meta_nrows, meta_ncols}));
    auto meta_reordered_cpu_ref =
      cutlass::TensorRef<MetaElement, MetaReorderedLayout>(
          meta_reordered_cpu.data_ptr<MetaElement>(),
          MetaReorderedLayout::packed({meta_nrows, meta_ncols}));
    reorder_meta(meta_reordered_cpu_ref, meta_cpu_ref, meta_nrows, meta_ncols);
  } else if (meta_dtype == at::kInt) {
    using MetaElement = int32_t;
    auto meta_cpu_ref =
      cutlass::TensorRef<MetaElement, MetaLayout>(
          meta_cpu.data_ptr<MetaElement>(),
          MetaLayout::packed({meta_nrows, meta_ncols}));
    auto meta_reordered_cpu_ref =
      cutlass::TensorRef<MetaElement, MetaReorderedLayout>(
          meta_reordered_cpu.data_ptr<MetaElement>(),
          MetaReorderedLayout::packed({meta_nrows, meta_ncols}));
    reorder_meta(meta_reordered_cpu_ref, meta_cpu_ref, meta_nrows, meta_ncols);
  }

  return std::make_tuple(sparse_cpu.to(dense.device()),
                         meta_reordered_cpu.to(dense.device()));
#else
  AT_ERROR("_semi_structured_sparse_compress_2by4_cutlass: ROCm doesn't support"
           "CUTLASS");
  return std::make_tuple(Tensor{}, Tensor{});
#endif
}

}  // namespace native
}  // namespace at
```
</details>

<details>
<summary>
Apply this patch on top of this PR
</summary>

```diff --git a/aten/src/ATen/native/native_functions.yaml b/aten/src/ATen/native/native_functions.yaml
index 1f07b6474bc..40f8cf3a6de 100644
--- a/aten/src/ATen/native/native_functions.yaml
+++ b/aten/src/ATen/native/native_functions.yaml
@@ -14949,3 +14949,8 @@
 # This op is ONLY used by pytorch/XLA in functionalization, and should never show up in vanilla eager mode or in any pytorch tracing contexts.
 - func: _propagate_xla_data(Tensor input, Tensor output) -> ()
   variants: function
+
+- func: _semi_structured_sparse_compress_2by4_cutlass(Tensor dense) -> (Tensor, Tensor)
+  variants: function
+  dispatch:
+    CPU, CUDA: _semi_structured_sparse_compress_2by4_cutlass
diff --git a/test/test_sparse.py b/test/test_sparse.py
index 5e8c29d671a..990f9b1c528 100644
--- a/test/test_sparse.py
+++ b/test/test_sparse.py
@@ -4932,6 +4932,10 @@ class TestSparseAny(TestCase):
 
             torch.testing.assert_close(dense, dense_ref, rtol=0, atol=0)
 
+            sparse_ref, meta_ref = torch.ops.aten._semi_structured_sparse_compress_2by4_cutlass(dense)
+            torch.testing.assert_close(sparse, sparse_ref, rtol=0, atol=0)
+            torch.testing.assert_close(meta, meta_ref, rtol=0, atol=0)
+
         for (m, k) in itertools.product(range(4), range(4)):
             m = (m + 1) * 32
             k = (k + 1) * 128

```
</details>

CUTLASS implementation of these operations is CPU only.  There is also a variation of the same code rewritten in CUDA in `aten/src/ATen/native/sparse/cuda/StructuredSparseLinearCUTLASS.cu`, that could be factored out for the same purpose.  However, as mentioned in the initial comment, this PR is not to be merged as is, but integrated with other ongoing work on semi-structured sparse tensor support.  For this reason, this type of test won't be incorporated right now.

Reply:
That might be a bit of heavy lifting for an automated test. Could we instead manufacture an input with a particular sparsity pattern and check the compression results against reference objects generated once ahead of time?

Reply:
> That might be a bit of heavy lifting for an automated test. Could we instead manufacture an input with a particular sparsity pattern and check the compression results against reference objects generated once ahead of time?

Yes, that would be a good idea.

Reply:
Added a comparison with CUTLASS-calculated metadata values into the test.